File: lambda/teams_formatter.py
"""Teams message formatting and adaptive card creation."""

import logging
from typing import Dict, Any
try:
    from .event_parser import ParsedEvent
except ImportError:
    from event_parser import ParsedEvent

logger = logging.getLogger(__name__)


def get_event_color(event_category: str, action: str, metadata: Dict[str, Any]) -> str:
    """
    Get theme color based on event type and action.
    
    Args:
        event_category: The event category (pull_request, push, comment, etc.)
        action: The specific action (created, merged, failed, etc.)
        metadata: Event metadata for additional context
    
    Returns:
        str: Hex color code for the event
    """
    # Red for failures
    if action in ['failed', 'declined', 'stopped'] or action == 'rejected':
        return "#DC3545"
    
    # Green for success actions
    if action in ['merged', 'succeeded', 'approved']:
        return "#28A745"
    
    # Check for failure states in commit status
    if event_category == 'commit_status':
        state = metadata.get('state', '').upper()
        if state in ['FAILED', 'STOPPED', 'ERROR']:
            return "#DC3545"
        elif state == 'SUCCESSFUL':
            return "#28A745"  # Green for success
        else:
            return "#FFC107"  # Yellow for in-progress
    
    # Blue for pull requests
    if event_category == 'pull_request':
        return "#0078D4"
    
    # Purple for push events
    if event_category == 'push':
        return "#6264A7"
    
    # Gray for comments and other events
    return "#6C757D"


def create_adaptive_card_data(parsed_event: ParsedEvent) -> Dict[str, Any]:
    """
    Create data payload for Adaptive Card template.
    
    Args:
        parsed_event: The parsed Bitbucket event
    
    Returns:
        dict: Data payload for Teams Workflow Adaptive Card template
    """
    # Base data that all events have
    data = {
        "title": parsed_event.title or f"{parsed_event.action.title()} in {parsed_event.repository}",
        "subtitle": f"by {parsed_event.author}" if parsed_event.author else None,
        "repository": parsed_event.repository,
        "action": parsed_event.action.title(),
        "author": parsed_event.author,
        "event_category": parsed_event.event_category,
        "description": parsed_event.description,
        "url": parsed_event.url
    }
    
    # Add event-specific data
    if parsed_event.event_category == 'pull_request':
        data.update({
            "pr_id": str(parsed_event.metadata.get('pr_id', '')),
            "source_branch": parsed_event.metadata.get('source_branch', 'unknown'),
            "target_branch": parsed_event.metadata.get('target_branch', 'unknown'),
            "state": parsed_event.metadata.get('state', 'unknown')
        })
    
    elif parsed_event.event_category == 'push':
        data.update({
            "branch": parsed_event.metadata.get('branch', 'unknown'),
            "commit_count": str(parsed_event.metadata.get('commit_count', 0)),
            "commits": parsed_event.metadata.get('commits', [])
        })
    
    elif parsed_event.event_category == 'comment':
        data.update({
            "context_title": parsed_event.metadata.get('context_title', 'unknown')
        })
    
    elif parsed_event.event_category == 'commit_status':
        data.update({
            "build_name": parsed_event.metadata.get('build_name', 'Build'),
            "build_status": parsed_event.metadata.get('state', 'unknown'),
            "commit_hash": parsed_event.metadata.get('commit_hash', 'unknown'),
            "branch": parsed_event.metadata.get('branch', 'unknown')
        })
    
    return data


def _get_text_color_for_theme(theme_color: str) -> str:
    """
    Get appropriate text color based on theme color.
    
    Args:
        theme_color: Hex color code
    
    Returns:
        str: Text color for Adaptive Card
    """
    # Map theme colors to text colors for better visibility
    color_map = {
        "#DC3545": "Attention",  # Red for failures
        "#28A745": "Good",       # Green for success
        "#FFC107": "Warning",    # Yellow for in-progress
        "#0078D4": "Accent",     # Blue for pull requests
        "#6264A7": "Default",    # Purple for push events
        "#6C757D": "Default"     # Gray for comments
    }
    
    return color_map.get(theme_color, "Default")


def format_teams_message(parsed_event: ParsedEvent) -> Dict[str, Any]:
    """
    Convert ParsedEvent to data payload for Teams Workflow.
    
    Args:
        parsed_event: The parsed Bitbucket event
    
    Returns:
        dict: Data payload for Teams Workflow Adaptive Card template
    
    Raises:
        ValueError: If parsed_event is None or has invalid data
    """
    if parsed_event is None:
        raise ValueError("ParsedEvent cannot be None")
    
    if not parsed_event.repository:
        raise ValueError("ParsedEvent must have a repository")
    
    # Create the data payload that the Teams Workflow template will use
    event_data = create_adaptive_card_data(parsed_event)
    
    # Add theme color for the template
    theme_color = get_event_color(parsed_event.event_category, parsed_event.action, parsed_event.metadata)
    event_data["theme_color"] = theme_color
    event_data["text_color"] = _get_text_color_for_theme(theme_color)
    
    # Add formatted commit details for push events
    if parsed_event.event_category == 'push':
        commits = event_data.get("commits", [])
        if commits:
            commit_details = []
            for commit in commits[:3]:  # Show max 3 commits
                commit_msg = commit.get('message', 'No message')[:50]
                if len(commit.get('message', '')) > 50:
                    commit_msg += '...'
                commit_details.append(f"• {commit.get('hash', 'unknown')}: {commit_msg}")
            event_data["commit_details"] = "\\n".join(commit_details)
    
    # Add branch flow for pull requests
    if parsed_event.event_category == 'pull_request':
        source = event_data.get('source_branch', 'unknown')
        target = event_data.get('target_branch', 'unknown')
        event_data["branch_flow"] = f"{source} → {target}"
    
    return event_data

File: src/teamsFormatter.ts
/**
 * Teams Message Formatting Module
 * Formats parsed Bitbucket events into Teams Adaptive Card data
 * Requirements: 7.1, 7.2, 7.3, 7.4, 7.5, 7.6, 7.7, 7.8
 */

export interface ParsedEvent {
  eventCategory: string;
  repository: string;
  action: string;
  author: string;
  title: string | null;
  description: string | null;
  url: string;
  metadata: Record<string, any>;
}

interface MentionEntity {
  type: string;
  text: string;
  mentioned: { id: string; name: string };
}

const COLOR_MAP: Record<string, string> = {
  failure: '#DC3545',
  success: '#28A745',
  'in-progress': '#FFC107',
  pull_request: '#0078D4',
  push: '#6264A7',
  default: '#6C757D'
};

const TEXT_COLOR_MAP: Record<string, string> = {
  '#DC3545': 'Attention',
  '#28A745': 'Good',
  '#FFC107': 'Warning',
  '#0078D4': 'Accent',
  '#6264A7': 'Default',
  '#6C757D': 'Default'
};

export function getEventColor(
  eventCategory: string,
  action: string,
  metadata: Record<string, any>
): string {
  if (['failed', 'declined', 'stopped', 'rejected'].includes(action)) {
    return COLOR_MAP.failure;
  }

  if (['merged', 'succeeded', 'approved'].includes(action)) {
    return COLOR_MAP.success;
  }

  if (eventCategory === 'commit_status') {
    const state = (metadata.state || '').toUpperCase();
    if (['FAILED', 'STOPPED', 'ERROR'].includes(state)) return COLOR_MAP.failure;
    if (state === 'SUCCESSFUL') return COLOR_MAP.success;
    return COLOR_MAP['in-progress'];
  }

  return COLOR_MAP[eventCategory] || COLOR_MAP.default;
}

export function createMentionEntity(email: string, displayName: string): MentionEntity | null {
  return email
    ? {
        type: 'mention',
        text: `<at>${displayName}</at>`,
        mentioned: { id: email, name: displayName }
      }
    : null;
}

export function createAdaptiveCardData(parsedEvent: ParsedEvent): Record<string, any> {
  const authorEmail = parsedEvent.metadata.author_email || '';
  const mentionEntity = createMentionEntity(authorEmail, parsedEvent.author);

  const data: Record<string, any> = {
    title:
      parsedEvent.title ||
      `${parsedEvent.action.charAt(0).toUpperCase() + parsedEvent.action.slice(1)} in ${parsedEvent.repository}`,
    subtitle: mentionEntity
      ? `by <at>${parsedEvent.author}</at>`
      : parsedEvent.author
        ? `by ${parsedEvent.author}`
        : null,
    repository: parsedEvent.repository,
    action: parsedEvent.action.charAt(0).toUpperCase() + parsedEvent.action.slice(1),
    author: parsedEvent.author,
    author_mention: mentionEntity ? `<at>${parsedEvent.author}</at>` : parsedEvent.author,
    event_category: parsedEvent.eventCategory,
    description: parsedEvent.description,
    url: parsedEvent.url,
    mention_entities: mentionEntity ? [mentionEntity] : []
  };

  // Add event-specific data
  if (parsedEvent.eventCategory === 'pull_request') {
    Object.assign(data, {
      pr_id: String(parsedEvent.metadata.pr_id || ''),
      source_branch: parsedEvent.metadata.source_branch || 'unknown',
      target_branch: parsedEvent.metadata.target_branch || 'unknown',
      state: parsedEvent.metadata.state || 'unknown'
    });
  } else if (parsedEvent.eventCategory === 'push') {
    Object.assign(data, {
      branch: parsedEvent.metadata.branch || 'unknown',
      commit_count: String(parsedEvent.metadata.commit_count || 0),
      commits: parsedEvent.metadata.commits || []
    });
  } else if (parsedEvent.eventCategory === 'comment') {
    data.context_title = parsedEvent.metadata.context_title || 'unknown';
  } else if (parsedEvent.eventCategory === 'commit_status') {
    Object.assign(data, {
      build_name: parsedEvent.metadata.build_name || 'Build',
      build_status: parsedEvent.metadata.state || 'unknown',
      commit_hash: parsedEvent.metadata.commit_hash || 'unknown',
      branch: parsedEvent.metadata.branch || 'unknown'
    });
  }

  return data;
}

export function formatTeamsMessage(parsedEvent: ParsedEvent): Record<string, any> {
  if (!parsedEvent || !parsedEvent.repository) {
    throw new Error(parsedEvent ? 'ParsedEvent must have a repository' : 'ParsedEvent cannot be null');
  }

  const eventData = createAdaptiveCardData(parsedEvent);
  const themeColor = getEventColor(parsedEvent.eventCategory, parsedEvent.action, parsedEvent.metadata);

  Object.assign(eventData, {
    theme_color: themeColor,
    text_color: TEXT_COLOR_MAP[themeColor] || 'Default'
  });

  // Add formatted commit details for push events
  if (parsedEvent.eventCategory === 'push') {
    const commits = eventData.commits || [];
    if (commits.length > 0) {
      eventData.commit_details = commits
        .slice(0, 3)
        .map((c: Record<string, any>) => {
          let msg = (c.message || 'No message').substring(0, 50);
          if ((c.message || '').length > 50) msg += '...';
          return `• ${c.hash || 'unknown'}: ${msg}`;
        })
        .join('\n');
    }
  }

  // Add branch flow for pull requests
  if (parsedEvent.eventCategory === 'pull_request') {
    eventData.branch_flow = `${eventData.source_branch || 'unknown'} → ${eventData.target_branch || 'unknown'}`;
  }

  eventData.mention_entities = (eventData.mention_entities || []).filter(
    (e: MentionEntity | null) => e !== null
  );

  return eventData;
}
